### PR TITLE
sql: implement basic SHOW HELP support

### PR DESCRIPTION
**Pull request by dt**
Only handles exact name matches at the moment.
Returns all overloads for the named function as well as their signature and category.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/9517)

### CODE REVIEW

**Comment by knz** [comment]
The implementation is clean and straightforward, kudos for that.
Although I do wonder about the wisdom of a special cased SHOW HELP statement. I think we would increase reusability of the help texts much more strongly and widely by having the help texts reachable through a virtual table, and have SHOW HELP simply an alias for `SELECT * FROM help WHERE name=XXX`. A virtual table buys us approximate lookups (`select * from help where name like 'set%'`), the ability to load the texts via SQL viewers, the ability to implement a help browser with a straightforward architecture, etc.
I get that you wanted to get a first version of this feature out, but perhaps we could file an issue already for a followup.

Let me know what you think.
:lgtm: otherwise!

---

Reviewed 6 of 6 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[sql/show.go, line 522 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9517#-KSUACVo0nZFI_bkuawz:-KSUACVpYTiMEw-dBAsZ:b-sqnmjb) ([raw file](https://github.com/cockroachdb/cockroach/blob/da21d72988c7a9c018c79b56f4e169bf22124c18/sql/show.go#L522)):*
> ```Go
> 
> 	name := strings.ToLower(n.Name.String())
> 	matches, ok := parser.Builtins[name]
> ```

I'd suggest a "TODO" here for searching alternative words if not match is found ("closest matches" or something similar).

---

*[sql/parser/builtins.go, line 94 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9517#-KSUAQ1aH11lG0oQZ0UX:-KSUAQ1aH11lG0oQZ0UY:bcogaol) ([raw file](https://github.com/cockroachdb/cockroach/blob/da21d72988c7a9c018c79b56f4e169bf22124c18/sql/parser/builtins.go#L94)):*
> ```Go
> 	impure         bool
> 	class          FunctionClass
> 	category, Info string
> ```

Since Category() is a method I think we should have Info() as a method too for consistency (unless you have a better argument for what you have written already).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9517#-:-KSUAxDTo4gMVoclV-Je:b-61y80z)*

File: sql/parser/show.go [hunk 1]
```diff
@@ -55,6 +55,17 @@ func (node *ShowDatabases) Format(buf *bytes.Buffer, f FmtFlags) {
 	buf.WriteString("SHOW DATABASES")
 }
 
+// ShowHelp represents a SHOW CONSTRAINTS statement.
```
Comment (bdarnell) [inline sql/parser/show.go]: s/CONSTRAINTS/HELP/

**Review by bdarnell — COMMENTED** [review]
Could we make `HELP` a top-level command instead of putting it under `SHOW`?

**Comment by dt** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[sql/parser/builtins.go, line 94 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9517#-KSUAQ1aH11lG0oQZ0UX:-KSWiEFpX5OaXc53C3d-:b-ogts9w) ([raw file](https://github.com/cockroachdb/cockroach/blob/da21d72988c7a9c018c79b56f4e169bf22124c18/sql/parser/builtins.go#L94)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Since Category() is a method I think we should have Info() as a method too for consistency (unless you have a better argument for what you have written already).
</details>
`Category` is a method since useful defaults are programmatically derivable (e.g. from the input and output types). `Info` OTOH doesn't have a similar usable default, so it seemed to make sense to just leave it a field.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9517)*

**Comment by dt** [comment]
Review status: 2 of 11 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[sql/show.go, line 522 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9517#-KSUACVo0nZFI_bkuawz:-KSbh5PhvyeAhwvdbbDT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/da21d72988c7a9c018c79b56f4e169bf22124c18/sql/show.go#L522)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> I'd suggest a "TODO" here for searching alternative words if not match is found ("closest matches" or something similar).
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9517)*

File: sql/parser/show.go (hunk 1 above)
Comment (dt) [inline sql/parser/show.go]: Done

**Comment by dt** [comment]
@knz I think I'd like to add cli usage and get a feel for it before refactoring to use virtual tables or whatever just since I think that may end up being instructive as to where we want the impl to go (and I'd I want to investigate some refactoring of virtual tables to use views that might make them a little cheaper to work with before adding more of them).

**Comment by knz** [comment]
Reviewed 5 of 10 files at r3, 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[sql/parser/builtins.go, line 94 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9517#-KSUAQ1aH11lG0oQZ0UX:-KSfH1GPRvN-9foKgO-v:b-7aqk5c) ([raw file](https://github.com/cockroachdb/cockroach/blob/da21d72988c7a9c018c79b56f4e169bf22124c18/sql/parser/builtins.go#L94)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary>
> `Category` is a method since useful defaults are programmatically derivable (e.g. from the input and output types). `Info` OTOH doesn't have a similar usable default, so it seemed to make sense to just leave it a field.
</details>
Conversely, having it as a field means we can't assert it was properly initialized. Also, having a single exported field in a struct where everything is non-exported otherwise really looks asymmetrical to me.

But I'll defer to your judgement on this. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9517)*

**Comment by knz** [comment]
I think this one is good to go!

**Pull request merged**